File: migrate_all.dart
import 'dart:io';

void main() {
  final files = [
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/core/shared/providers/settings_provider.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/core/shared/providers/theme_provider.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/downloads/presentation/downloads_screen.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/library/presentation/library_screen.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/player/presentation/player_provider.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/search/presentation/search_provider.dart"
  ];

  for (final path in files) {
    final file = File(path);
    if (!file.existsSync()) continue;

    String content = file.readAsStringSync();

    // Generic replace
    content = content.replaceAll("StateNotifierProvider", "NotifierProvider");
    content = content.replaceAll("StateNotifier", "Notifier");

    // Pattern 1: settings_provider init extraction
    final pattern1 = RegExp(r"final\s+(\w+)\s*=\s*NotifierProvider<(\w+),\s*([^>]+)>\(\(ref\)\s*\{\s*([\s\S]*?)return\s+\w+\([^)]*\);\s*\}\);\s*class\s+\2\s+extends\s+Notifier<\3>\s*\{\s*\2\(super\.state\);");
    content = content.replaceAllMapped(pattern1, (match) {
      final providerName = match.group(1)!;
      final name = match.group(2)!;
      final type = match.group(3)!;
      final initLogic = match.group(4)!;
      
      return "final $providerName =\n    NotifierProvider<$name, $type>($name.new);\n\nclass $name extends Notifier<$type> {\n  @override\n  $type build() {\n    ${initLogic.trim()}\n  }";
    });

    // Pattern 2: Empty super constructor
    final pattern2 = RegExp(r"class\s+(\w+)\s+extends\s+Notifier<([^>]+)>\s*\{\s*\1\(\)\s*:\s*super\((.*?)\);");
    content = content.replaceAllMapped(pattern2, (match) {
      final name = match.group(1)!;
      final type = match.group(2)!;
      final init = match.group(3)!;
      return "class $name extends Notifier<$type> {\n  @override\n  $type build() => $init;";
    });

    // Pattern 3: Provider init for empty constructors
    final pattern3 = RegExp(r"final\s+(\w+)\s*=\s*NotifierProvider<(\w+),\s*([^>]+)>\(\s*\(ref\)\s*=>\s*\2\(\)\s*\);");
    content = content.replaceAllMapped(pattern3, (match) {
      final providerName = match.group(1)!;
      final name = match.group(2)!;
      final type = match.group(3)!;
      return "final $providerName = NotifierProvider<$name, $type>($name.new);";
    });

    file.writeAsStringSync(content);
    print("Migrated $path");
  }
}

File: migrate_all.py
import re
import os

files_to_migrate = [
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/core/shared/providers/settings_provider.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/core/shared/providers/theme_provider.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/downloads/presentation/downloads_screen.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/library/presentation/library_screen.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/player/presentation/player_provider.dart",
    "e:/dinu/Learn/flutter-learn/Sonexa/lib/features/search/presentation/search_provider.dart"
]

def migrate_file(path):
    with open(path, "r", encoding="utf-8") as f:
        content = f.read()

    # Generic replace
    content = content.replace("StateNotifierProvider", "NotifierProvider")
    content = content.replace("StateNotifier", "Notifier")

    # Simple provider init extraction (like in settings_provider)
    pattern1 = r"final\s+(\w+)\s*=\s*NotifierProvider<(\w+),\s*([^>]+)>\(\(ref\)\s*\{\s*([\s\S]*?)return\s+\w+\([^)]*\);\s*\}\);\s*class\s+\2\s+extends\s+Notifier<\3>\s*\{\s*\2\(super\.state\);"
    def replacer1(match):
        provider_name = match.group(1)
        name = match.group(2)
        type_ = match.group(3)
        init_logic = match.group(4)
        
        provider_code = f"final {provider_name} =\n    NotifierProvider<{name}, {type_}>({name}.new);"
        notifier_code = f"class {name} extends Notifier<{type_}> {{\n  @override\n  {type_} build() {{\n    {init_logic.strip()}\n  }}"
        return provider_code + "\n\n" + notifier_code

    content = re.sub(pattern1, replacer1, content)
    
    # Empty super constructor without state (e.g. MyNotifier() : super(init);)
    # class NameNotifier extends Notifier<Type> {
    #   NameNotifier() : super(init);
    pattern2 = r"class\s+(\w+)\s+extends\s+Notifier<([^>]+)>\s*\{\s*\1\(\)\s*:\s*super\((.*?)\);"
    def replacer2(match):
        name = match.group(1)
        type_ = match.group(2)
        init = match.group(3)
        return f"class {name} extends Notifier<{type_}> {{\n  @override\n  {type_} build() => {init};"
    
    content = re.sub(pattern2, replacer2, content)

    # Provider init for empty constructors
    # final nameProvider = NotifierProvider<NameNotifier, Type>((ref) => NameNotifier());
    pattern3 = r"final\s+(\w+)\s*=\s*NotifierProvider<(\w+),\s*([^>]+)>\(\s*\(ref\)\s*=>\s*\2\(\)\s*\);"
    def replacer3(match):
        provider_name = match.group(1)
        name = match.group(2)
        type_ = match.group(3)
        return f"final {provider_name} = NotifierProvider<{name}, {type_}>({name}.new);"
    
    content = re.sub(pattern3, replacer3, content)

    with open(path, "w", encoding="utf-8") as f:
        f.write(content)
    print(f"Migrated {os.path.basename(path)}")

for f in files_to_migrate:
    migrate_file(f)
